Add custom unbaked block state model types

PepperCode1 opened the pull request:
This pull request adds the `CustomUnbakedBlockStateModel`, which allows using custom model types in blockstate files. Any place where a `ModelVariant` is normally allowed (including within a weighted list) can now use the `"fabric:type"` key and define a custom model.

The `fabric:composite` custom model type is registered by default. It simply combines the geometry from multiple submodels and is analogous to vanilla's `minecraft:composite` item model type. Submodels are defined using the `"models"` key, whose type is an array of arbitrary models.

### TODO

- [x] Consider moving the `CompositeBlockStateModel` into the API package, or at least expose static methods to construct the baked and unbaked instances.
  **Resolution:** Added `CompositeBlockStateModel` interface to API that exposes construction and inspection for both baked and unbaked types. Renamed implementation class.
- [x] Consider adding a default custom type that works like `ModelVariant` but accepts a much more flexible transform that allows rotation about the Z axis, non-90 degree rotations, scaling, and translation.
  **Resolution:** Will not be added in this PR due to exact format needing additional discussion and functional overlap with static model root transforms, which may be added in the future.
- [x] Documentation

PepperCode1 committed:
Add custom unbaked block state model types

PepperCode1 committed:
Add documentation

Technici4n commented on fabric-model-loading-api-v1/src/client/java/net/fabricmc/fabric/impl/client/model/loading/CompositeBlockStateModel.java:
```diff
@@ -0,0 +1,141 @@
+				subkeys[i] = subkey;
+			}
+
+			record Key(Object[] subkeys) {
```
Are you sure that this will produce suitable `equals` and `hashCode`?

PepperCode1 replied:
I don't see why not; it is almost exactly the same code as is used by the multipart model.

PepperCode1 replied:
It turns out arrays do not override `equals` and `hashCode` for whatever reason. I will replace the array with a `List`.

PepperCode1 committed:
Fix composite and multipart models' createGeometryKey implementation

Arrays do not implement equals and hashCode, so use an ArrayList instead, which has the most efficient equals and hashCode implementations of any available list implementation.

modmuss50 reviewed (approved):
Looks great, many thanks :)

PepperCode1 committed:
Expose composite model in API

PepperCode1 commented:
Another WIP PR depends on this one, so this PR has been merged early to a staging branch. Any changes can still be made before a full release, so feel free to review it.
